unison-src/new-runtime-transcripts/hashing.output.md

```diff
@@ -137,11 +137,11 @@ ex5 = crypto.hmac Sha2_256 mysecret f |> hex
   
     25 | > ex4
            ⧩
-           "30eb272078faa05fec07601353cb007c90977778965bb31d0c097c7191deb22f"
+           "43cccb245b4b50964e71dff9a53c330f38320673252ab9110e4bac8c3a965354"
   
     26 | > ex5
            ⧩
-           "978e69198db0ff7b2b17a976a681eceae26d5ed5642f96987b5ee7b10ca01274"
+           "8f8785a7442b8363df1f2ed560f6ea42132a63824984b531a23187ee94c110c4"
```

Comment:
Why?

Reply:
Would like to understand this one.

Reply:
Okay, this isn't affecting term or type hashing since that would have changed other transcripts.

@dolio this looks to be a change to the output of universal hashing - it is hashing the identity function in both cases I believe. My guess is that there is something about the implementation of universal hashing that depends on some implementation detail of one of the libraries we are using, and that implementation detail changed between library versions. Any ideas on what that might be?

We want the hashing algorithm to be stable and not change unintentionally, so I'd like to get to the bottom of this.

Reply:
> I'd like to understand the hashing output before merging. Are any other hashes affected? Like the worst thing would be if this silently changes all the hashing we are doing with cryptonite everywhere, including for hashing terms and types, etc.

Agreed.

> Also there are a couple Mac CI targets failing, do you know what that is?

It's failing to build in Big Sur on GitHub CI; it doesn't even start the VM.
> "GitHub Actions has encountered an internal error when running your job."

It does, however, build with `stack` on Big Sur, so if we can find a satisfactory answer to the transcript/hash thing, I'd just revert the last two CI-related commits and deal with it in a separate ticket.

Reply:
The serialization output seems to be changing slightly every time I restart ucm with the watch cache disabled, so something funky is going on.

Reply:
Okay, I think I figured out what's going on. The translation of combinator references from one level to another is just like `f (CIx r m n) = GR r m n`. However, this makes no sense, because one of those numbers (`m`) corresponds to the locally allocated number for the combinator. So it should be thrown out when serializing and figured out from the reference when deserializing.

Reply:
@dolio Good sleuthing. So basically, the issue is present in trunk, it's just that you wouldn't notice it unless you rerun the transcripts with the watch cache disabled.

I was trying to figure out why CI passes but this might be the first PR merged since the universal hashing landed. Also possible that the transcript runner is still deterministic since even if the hashing depends on implementation details, it is always running the transcripts in the exact same order.

In any case, @dolio can you create an issue to track and @aryairani I think it is fine to merge this now.

Reply:
Also I thought of an idea for when you fix this @dolio - randomize the order that transcripts are run in the test harness - I'm pretty sure it just takes a directory listing so you can just shuffle that before iterating over it. It shouldn't affect any of the transcript output and might catch some issues like this.

Reply:
Will that do anything? I thought every transcript was run in its own execution, essentially, so they're all in separate runtime instances. Randomizing the order would only matter for this if multiple transcripts were getting loaded in a single runtime environment.

I'm not really sure why we get nondeterminstic results here, even. I guess a bunch of definitions from a file can end up represented in slightly different orders or something.

Reply:
Oh... I was under the impression that the same runtime instance gets recycled for each of the transcripts, but I could be wrong about that.

Reply:
I think the transcript tests just system call out to `unison transcript whatever.md` for each file.